### ISSUE TITLE
MPI B&S does not work correctly with "hanging" local nodes

### ISSUE THREAD

**Issue by ddiezrod** [KratosMultiphysics/Kratos#10994]
I have found that the global system in trilinos block builder and solver does not work correctly if we have a partition with a hanging local node, i.e. a local node that belongs to an element that is in a different partition. If the node does not belong to any element in its own partition then it is not considered in the global system here
https://github.com/KratosMultiphysics/Kratos/blob/b1958c394fe5d7b99d3c8e807fe1c05c799d4ff3/applications/TrilinosApplication/custom_strategies/builder_and_solvers/trilinos_block_builder_and_solver.h#L569

In principle, metis makes sure that this never happens, if there is a hanging node it should be ghost, not local. The problem is that in my case the computing model part is a submodelpart from the main model part that is used in the partitioning, so this can happen.

I understand that the way to correct this would be to synchronize dofs so these hanging nodes are also added to mDofSet somewhere here
https://github.com/KratosMultiphysics/Kratos/blob/b1958c394fe5d7b99d3c8e807fe1c05c799d4ff3/applications/TrilinosApplication/custom_strategies/builder_and_solvers/trilinos_block_builder_and_solver.h#L518

I am not sure how this should be done though, or if this could break other parts of the code or affect performance significantly.

Ping @KratosMultiphysics/technical-committee @KratosMultiphysics/implementation-committee @jcotela @loumalouomega  @philbucher

**Comment by jcotela** [comment]
Added a failing test on this branch https://github.com/KratosMultiphysics/Kratos/tree/mpi/builder-and-solver-hanging-node-support

**Comment by loumalouomega** [comment]
> Added a failing test on this branch [`mpi`/builder-and-solver-hanging-node-support](https://github.com/KratosMultiphysics/Kratos/tree/mpi/builder-and-solver-hanging-node-support?rgh-link-date=2023-04-15T16%3A44%3A00Z)

Thanks, I have been thinking about this, and I think we can discuss a solution on Monday. The drawback is that we will need a new flag to register, and I think we are almost out of flags to register.

**Comment by jcotela** [comment]
Actually, I just saw that my test actually fails on the ParallelFillCommunicator, it does not even reach the B&S... We may want to fix that too, but it is a separate issue.

I will update the test to more closely match the situation described by @ddiezrod: He has a well-formed model part that ParallelFillCommunicator happily processes, but then works on a "bad" sub-model part of that, I am creating a bad modelpart from the start here, which fails earlier.

We talk about it on Monday.

**Comment by loumalouomega** [comment]
> Actually, I just saw that my test actually fails on the ParallelFillCommunicator, it does not even reach the B&S... We may want to fix that too, but it is a separate issue.
> 
> I will update the test to more closely match the situation described by @ddiezrod: He has a well-formed model part that ParallelFillCommunicator happily processes, but then works on a "bad" sub-model part of that, I am creating a bad modelpart from the start here, which fails earlier.
> 
> We talk about it on Monday.

In fact, part of my idea was to modify the ParallelFillCommunicator, see it on Monday

**Comment by jcotela** [comment]
I updated the branch with a fix, now the test passes. I also slightly modified the test to make it more accurate/generic: now the partition without elements has two active nodes without an element (one local and one ghost) and a third, completely unused node (which should not get dofs)

As far as I can tell, the solution in the branch is fully functional (it solves the open issue), but the extra synchronization step (including MPI communication and an extra sort+unique of the DofSet) is always active. To conclude the discussion, we should decide if we want to be able to toggle this (probably leaving it off by default to avoid the extra cost, which is not needed in most cases)

**Comment by loumalouomega** [comment]
As I said, I am okay of this if not defaulted, because that extra cost
could be problematic. In order to do so, Parameters constructor should be
consider to avoid populate with many constructors. I added in a different
PR an initial consideration of the Parameters in the constructor.

**Comment by ddiezrod** [comment]
@jcotela Can you create a draft PR so we can comment the changes directly there?

**Comment by loumalouomega** [comment]
> @jcotela Can you create a draft PR so we can comment the changes directly there?

Indeed, may be interesting to review it

**Comment by jcotela** [comment]
PR in #11059. For now I have not made it dependent on #10887, since I believe that the changes are largely orthogonal to that PR.

**Comment by RiccardoRossi** [comment]
but still, this is not something that belongs to the builder and solver....dofs should be in sync outside of the builder and solver, so it should be fixed elsewhere
